### PR TITLE
RTBHouse: native ads support

### PR DESCRIPTION
**Pull request by piotrj-rtbh**
## Type of change
[x] Feature

## Description of change
Native ads are now available to be processed and accessed by RTB House adapter.

## Other information
Please reach us at [inventory_support@rtbhouse.com](mailto:inventory_support@rtbhouse.com) with [piotr.jaworski@rtbhouse.com](mailto:piotr.jaworski@rtbhouse.com) and cc [leandro.otani@rtbhouse.com](mailto:leandro.otani@rtbhouse.com).

### CODE REVIEW

File: src/main/java/org/prebid/server/bidder/rtbhouse/RtbhouseBidder.java [hunk 1]
```diff
@@ -83,14 +86,16 @@ public Result<List<HttpRequest<BidRequest>>> makeHttpRequests(BidRequest bidRequ
     @Override
     public Result<List<BidderBid>> makeBids(BidderCall<BidRequest> httpCall, BidRequest bidRequest) {
         try {
+            final List<BidderError> bidderErrors = new ArrayList<>();
             final BidResponse bidResponse = mapper.decodeValue(httpCall.getResponse().getBody(), BidResponse.class);
-            return Result.withValues(extractBids(bidResponse));
-        } catch (DecodeException e) {
+            return Result.of(extractBids(bidRequest, bidResponse, bidderErrors), Collections.emptyList());
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/rtbhouse/RtbhouseBidder.java:91]: Please, use `Result.withValues` here.

File: src/main/java/org/prebid/server/bidder/rtbhouse/RtbhouseBidder.java [hunk 2]
```diff
@@ -83,14 +86,16 @@ public Result<List<HttpRequest<BidRequest>>> makeHttpRequests(BidRequest bidRequ
...
+        } catch (DecodeException | PreBidException e) {
             return Result.withError(BidderError.badServerResponse(e.getMessage()));
         }
     }
 
-    private static List<BidderBid> extractBids(BidResponse bidResponse) {
+    private List<BidderBid> extractBids(BidRequest bidRequest,
+            BidResponse bidResponse, List<BidderError> bidderErrors) {
         if (bidResponse == null || CollectionUtils.isEmpty(bidResponse.getSeatbid())) {
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/rtbhouse/RtbhouseBidder.java:99]: Please, format this lines like this:
```
private List<BidderBid> extractBids(BidRequest bidRequest,
                                    BidResponse bidResponse,
                                    List<BidderError> bidderErrors) {

         if (bidResponse == null || CollectionUtils.isEmpty(bidResponse.getSeatbid())) {
```

File: src/main/java/org/prebid/server/bidder/rtbhouse/RtbhouseBidder.java [hunk 3]
```diff
@@ -101,10 +106,60 @@ private static List<BidderBid> extractBids(BidResponse bidResponse) {
...
                 .flatMap(Collection::stream)
                 .filter(Objects::nonNull)
-                .map(bid -> BidderBid.of(bid, BidType.banner, bidResponse.getCur()))
+                .map(bid -> resolveBidderBid(bid, bidRequest, bidResponse, bidderErrors))
                 .toList();
     }
 
+    private BidderBid resolveBidderBid(Bid bid, BidRequest bidRequest,
+                BidResponse bidResponse, List<BidderError> bidderErrors) {
+        final String currency = bidResponse.getCur();
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/rtbhouse/RtbhouseBidder.java:115]: Please, format this lines like in example above.

File: src/main/java/org/prebid/server/bidder/rtbhouse/RtbhouseBidder.java [hunk 4]
```diff
@@ -101,10 +106,60 @@ private static List<BidderBid> extractBids(BidResponse bidResponse) {
...
+        final JsonNode admNode;
+        try {
+            admNode = mapper.mapper().readTree(adm);
+        } catch (JsonProcessingException e) {
+            bidderErrors.add(BidderError.badServerResponse("Unable to parse native adm: %s".formatted(adm)));
+            return null;
+        }
+
+        final JsonNode nativeNode = admNode.get("native");
+        if (!nativeNode.isMissingNode()) {
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/rtbhouse/RtbhouseBidder.java:142]: Please, choose one of the solutions and use it here:
```
final JsonNode nativeNode = admNode.get("native");
if (nativeNode != null) {
```
or
```
private static final JsonPointer NATIVE_FIELD = JsonPointer.valueOf("/native");;
```
```
final JsonNode nativeNode = admNode.at(NATIVE_FIELD);
if (!nativeNode.isMissingNode()) {
```

File: src/main/java/org/prebid/server/bidder/rtbhouse/RtbhouseBidder.java [hunk 5]
```diff
@@ -101,10 +106,60 @@ private static List<BidderBid> extractBids(BidResponse bidResponse) {
...
+            if (imp.getId().equals(impId)) {
+                if (imp.getBanner() != null) {
+                    return bidType;
+                } else if (imp.getXNative() != null) {
+                    bidType = BidType.xNative;
+                }
+            }
+        }
+        return bidType;
+    }
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/rtbhouse/RtbhouseBidder.java:161]: Please, refactor this method like this:
```
    private static BidType getBidType(String impId, List<Imp> imps) {
        for (Imp imp : imps) {
            if (imp.getId().equals(impId)) {
                if (imp.getBanner() != null) {
                    return BidType.banner;
                } else if (imp.getXNative() != null) {
                    return BidType.xNative;
                }
            }
        }
        return BidType.banner;
    }
```

**Review by CTMBNara — CHANGES_REQUESTED** [review]

File: src/main/java/org/prebid/server/bidder/rtbhouse/RtbhouseBidder.java (hunk 1 above)
Comment (piotrj-rtbh) [reply src/main/java/org/prebid/server/bidder/rtbhouse/RtbhouseBidder.java]: Done

File: src/main/java/org/prebid/server/bidder/rtbhouse/RtbhouseBidder.java (hunk 2 above)
Comment (piotrj-rtbh) [reply src/main/java/org/prebid/server/bidder/rtbhouse/RtbhouseBidder.java]: Done

File: src/main/java/org/prebid/server/bidder/rtbhouse/RtbhouseBidder.java (hunk 4 above)
Comment (piotrj-rtbh) [reply src/main/java/org/prebid/server/bidder/rtbhouse/RtbhouseBidder.java]: Fixed using the first solution.

File: src/main/java/org/prebid/server/bidder/rtbhouse/RtbhouseBidder.java (hunk 5 above)
Comment (piotrj-rtbh) [reply src/main/java/org/prebid/server/bidder/rtbhouse/RtbhouseBidder.java]: Refactored, thanks.

File: src/main/java/org/prebid/server/bidder/rtbhouse/RtbhouseBidder.java (hunk 3 above)
Comment (piotrj-rtbh) [reply src/main/java/org/prebid/server/bidder/rtbhouse/RtbhouseBidder.java]: Done

**Pull request merged**